File: protocol-generator/python/src/entity_core/peer/transport.py
"""Transport (L4): TCP listener + thread-per-connection reader-demux (§1.6
framing, §4.8 inbound concurrency, §6.11 reentry) + the client dialer/handshake
that drives the two-peer loopback.

CONCURRENCY MODEL (profile [async].style = threaded, thread-per-connection):
one reader thread per connection demuxes inbound frames (§6.11).  An
EXECUTE_RESPONSE routes to its awaiting outbound caller by request_id through a
per-conn pending-map + a single :class:`threading.Condition` (the threaded-peer
reader-thread + condvar shape, same as the OCaml / Ruby cohort).  An inbound
EXECUTE is dispatched on ITS OWN thread (§4.8) so a handler that originates an
outbound EXECUTE (§6.11 reentry) and awaits its response does NOT block the
reader.  Writes are serialized by a Lock.  CPython releases the GIL during
blocking socket IO + inside the cryptography/hashlib C extensions, so this is
genuinely concurrent for the IO-bound peer workload (A-PY-007).

TCP_NODELAY (profile [async].tcp_nodelay = true): set on EVERY accepted/dialed
socket from day one — Nagle + delayed-ACK on small req/resp frames is THE §7b
throughput killer.
"""

from __future__ import annotations

import socket
import threading
from typing import Callable

from .model import Entity, Envelope
from .wire import (
    FrameTooLargeError,
    frame_of_envelope,
    make_execute,
    make_response,
    read_frame,
    response_result,
    response_status,
    write_frame,
)
from .peer import Conn, Peer
from .identity import Identity, peer_id_of_public_key


def _set_nodelay(sock: socket.socket) -> None:
    try:
        sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
    except OSError:
        pass


class TransportIO:
    """A per-connection IO endpoint shared by server + client."""

    def __init__(self, sock: socket.socket) -> None:
        _set_nodelay(sock)
        self.sock = sock
        self._write_lock = threading.Lock()
        self._cond = threading.Condition()
        self._pending: dict[str, Envelope | None] = {}  # request_id -> response
        self._waiting: set[str] = set()
        self._closed = False

    def write_framed(self, env: Envelope) -> None:
        payload = frame_of_envelope(env)
        with self._write_lock:
            write_frame(self.sock, payload)

    def _route_response(self, env: Envelope) -> None:
        request_id = env.root.text("request_id") or ""
        with self._cond:
            if request_id in self._waiting:
                self._pending[request_id] = env
                self._cond.notify_all()

    def outbound(self, req: Envelope) -> Envelope | None:
        """Send a request envelope and await its correlated EXECUTE_RESPONSE
        (§6.11 demux).  Returns None if the connection closes first."""
        request_id = req.root.text("request_id") or ""
        with self._cond:
            if self._closed:
                return None
            self._waiting.add(request_id)
        try:
            self.write_framed(req)
        except OSError:
            with self._cond:
                self._waiting.discard(request_id)
            return None
        with self._cond:
            while request_id not in self._pending and not self._closed:
                self._cond.wait()
            self._waiting.discard(request_id)
            return self._pending.pop(request_id, None)

    def close(self) -> None:
        with self._cond:
            if self._closed:
                return
            self._closed = True
            self._cond.notify_all()
        try:
            self.sock.close()
        except OSError:
            pass

    def _reject_non_canonical(self, payload: bytes) -> None:
        """Answer a frame the strict decoder rejected with ``400 non_canonical_ecf``
        (§6.3), recovering ONLY the ``request_id`` so the sender can correlate it.

        The frame stays rejected: nothing is built from it, nothing is stored, and the
        tag is never interpreted — the salvage decode exists solely to read back the
        correlation key. The envelope and entity-wrapper shapes are fixed maps with no
        legal tag position, so a frame whose ONLY defect is a tag inside some entity's
        ``data`` still has a structurally sound root, which is exactly the case worth
        recovering (and the one CAP-6a's ``>2^64`` half arrives as — a bignum can only
        reach a peer as a major-type-6 tag). If even the request_id is unrecoverable
        there is nobody to answer, so the frame is dropped: the one case where silence is
        all that is available.
        """
        from .._cbor import decode_salvage
        from .model import Envelope, Included
        from .wire import error_result, make_response

        try:
            v = decode_salvage(payload)
            request_id = v["root"]["data"]["request_id"]
            if not isinstance(request_id, str):
                return
        except Exception:
            return  # no correlatable request_id — nothing to answer
        try:
            result = error_result(
                "non_canonical_ecf",
                "frame is not canonical ECF (section 6.3): CBOR tags are forbidden "
                "anywhere in an entity",
            )
            self.write_framed(
                Envelope(root=make_response(request_id, 400, result), included=Included())
            )
        except Exception:
            # A write failure here is a dead socket, not a protocol decision; the read
            # loop's own error handling ends the loop on the next iteration.
            return

    def read_loop(self, on_execute: Callable[[Envelope], None]) -> None:
        """§6.11 demux: EXECUTE_RESPONSE -> route; EXECUTE -> dispatch on its own
        thread (§4.8)."""
        from .model import BadEntityError
        from .wire import envelope_of_frame

        while True:
            try:
                payload = read_frame(self.sock)
            except FrameTooLargeError:
                # §4.10(a): rejected before buffering; close + keep the peer
                # serving other connections (this loop just ends).
                return
            except (OSError, ConnectionError):
                return
            try:
                env = envelope_of_frame(payload)
            except (BadEntityError, Exception):
                # §6.3: "Rejection returns `400 non_canonical_ecf`" — the frame is
                # refused (correct), and that refusal MUST be a STATUS, not silence.
                # Skipping it satisfies only the first half of the sentence and leaves
                # the sender blocked until its own timeout, so a refusal is
                # indistinguishable from a dead peer. §4.9(c) deliver-or-signal says the
                # same from the other direction. Answer, then keep reading.
                self._reject_non_canonical(payload)
                continue
            if env.root.type == "system/protocol/execute/response":
                self._route_response(env)
            else:
                threading.Thread(target=on_execute, args=(env,), daemon=True).start()


# ── server ────────────────────────────────────────────────────────────────────
class Listener:
    """A running TCP listener for a peer."""

    def __init__(self, peer: Peer, port: int) -> None:
        self.peer = peer
        self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._sock.bind(("127.0.0.1", port))
        self._sock.listen(64)
        self.port = self._sock.getsockname()[1]
        self._accept_thread = threading.Thread(target=self._accept_loop, daemon=True)
        self._accept_thread.start()

    def _accept_loop(self) -> None:
        while True:
            try:
                conn_sock, _ = self._sock.accept()
            except OSError:
                return
            threading.Thread(target=self._serve, args=(conn_sock,), daemon=True).start()

    def _serve(self, conn_sock: socket.socket) -> None:
        tio = TransportIO(conn_sock)
        cn = Conn()
        cn.outbound = tio.outbound  # wire the §6.11 reentry seam to this conn

        def on_execute(env: Envelope) -> None:
            # per-request isolation: an adversarial request must NOT tear down the
            # connection (§4.9 no-crash; §3.3 every EXECUTE gets a response).
            try:
                resp = self.peer.dispatch(cn, env)
                if resp is not None:
                    tio.write_framed(resp)
            except Exception:
                request_id = env.root.text("request_id") or ""
                from .wire import error_result

                try:
                    tio.write_framed(Envelope.of(
                        make_response(request_id, 500, error_result("internal_error"))
                    ))
                except OSError:
                    pass

        tio.read_loop(on_execute)
        tio.close()

    def close(self) -> None:
        try:
            self._sock.close()
        except OSError:
            pass


def listen(peer: Peer, port: int = 0) -> Listener:
    """Bind 127.0.0.1:port (0 = auto-assign) and start accepting."""
    return Listener(peer, port)


# ══════════════════════════════════════════════════════════════════════════════
# Client side — dialer + initiator handshake (drives the two-peer loopback)
# ══════════════════════════════════════════════════════════════════════════════
class HandshakeError(Exception):
    def __init__(self, step: str, status: int = 0, code: str = "") -> None:
        super().__init__(f"{step} failed: status {status} {code}")
        self.step = step
        self.status = status
        self.code = code


class ClientConnection:
    """An initiator-side session that drives the §4.1 handshake + authenticated
    EXECUTEs over a real TCP connection."""

    def __init__(self, host: str, port: int) -> None:
        sock = socket.create_connection((host, port))
        self._io = TransportIO(sock)
        self._req_counter = 0
        self.remote_peer_id = ""
        self.capability: Entity | None = None
        self._granter_peer: Entity | None = None
        self._cap_sig: Entity | None = None
        threading.Thread(target=self._io.read_loop, args=(lambda env: None,), daemon=True).start()

    def _next_request_id(self) -> str:
        self._req_counter += 1
        return "req-" + str(self._req_counter)

    def _send(self, req: Envelope) -> Envelope | None:
        return self._io.outbound(req)

    def close(self) -> None:
        self._io.close()

    # ── initiator handshake (§4.1 forward leg: hello -> authenticate) ────────
    def handshake(self, local: Identity) -> None:
        import os

        from .model import Envelope as Env

        hello = Entity.make("system/protocol/connect/hello", {
            "peer_id": local.peer_id,
            "nonce": os.urandom(32),
            "protocols": ["entity-core/1.0"],
            "timestamp": int(__import__("time").time() * 1000),
            "hash_formats": ["ecfv1-sha256"],
            "key_types": ["ed25519"],
        })
        r1 = self._send(Env.of(make_execute(self._next_request_id(), "system/protocol/connect", "hello", hello)))
        if r1 is None:
            raise HandshakeError("hello", 0, "connection_broken")
        self._require_ok(r1, "hello")
        remote_hello = response_result(r1)
        if remote_hello is None:
            raise HandshakeError("hello", code="bad_response")
        self.remote_peer_id = remote_hello.text("peer_id") or ""
        remote_nonce = remote_hello.bytes_("nonce")

        auth = Entity.make("system/protocol/connect/authenticate", {
            "peer_id": local.peer_id,
            "public_key": bytes(local.public_key),
            "key_type": "ed25519",
            "nonce": remote_nonce,
        })
        auth_sig = local.sign_entity(auth)
        r2 = self._send(Env.of(
            make_execute(self._next_request_id(), "system/protocol/connect", "authenticate", auth),
            local.peer_entity,
            auth_sig,
        ))
        if r2 is None:
            raise HandshakeError("authenticate", 0, "connection_broken")
        self._require_ok(r2, "authenticate")
        grant = response_result(r2)
        if grant is None:
            raise HandshakeError("authenticate", code="bad_grant")
        token_h = grant.bytes_("token")
        token = r2.included.get_by_hash(token_h)
        if token is None:
            raise HandshakeError("authenticate", code="missing_token")
        granter_h = token.bytes_("granter")
        granter_peer = r2.included.get_by_hash(granter_h)
        cap_sig = _find_signature(token.hash, r2.included)
        if granter_peer is None or cap_sig is None:
            raise HandshakeError("authenticate", code="missing_grant_authority")
        self.capability = token
        self._granter_peer = granter_peer
        self._cap_sig = cap_sig

    def _require_ok(self, env: Envelope, step: str) -> None:
        status = response_status(env)
        if status == 200:
            return
        code = ""
        res = response_result(env)
        if res is not None:
            code = res.text("code") or ""
        raise HandshakeError(step, status, code)

    # ── authenticated EXECUTE (§5.8 full authority chain in `included`) ──────
    def execute(self, local: Identity, uri: str, operation: str, params: Entity, resource=None) -> Envelope | None:
        from .model import Envelope as Env

        if self.capability is None:
            raise RuntimeError("execute before handshake")
        exec_e = make_execute(
            self._next_request_id(), uri, operation, params,
            author=local.identity_hash,
            capability=self.capability.hash,
            resource=resource,
        )
        exec_sig = local.sign_entity(exec_e)
        env = Env.of(
            exec_e, self.capability, self._granter_peer, local.peer_entity, self._cap_sig, exec_sig
        )
        return self._send(env)


def _find_signature(target: bytes, included) -> Entity | None:
    from .capability import find_signature

    return find_signature(target, included)


def dial(host: str, port: int) -> ClientConnection:
    return ClientConnection(host, port)

File: protocol-generator/kotlin/src/main/kotlin/org/entitycore/protocol/peer/Transport.kt
package org.entitycore.protocol.peer

import org.entitycore.protocol.EcfResult
import org.entitycore.protocol.EntityError
import org.entitycore.protocol.codec.CanonicalCbor
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeoutOrNull
import org.entitycore.protocol.codec.EcfValue
import java.io.BufferedOutputStream
import java.io.DataInputStream
import java.io.IOException
import java.io.OutputStream
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import kotlin.concurrent.thread

/**
 * Transport (L4): TCP listener + dialer, per-connection reader, §6.11 request_id demux,
 * the §4.8 inbound-concurrent-with-outbound dispatch, and the §6.13(b) reentry seam.
 * Plus the initiator dialer/handshake that drives the loopback.
 *
 * **Concurrency model (A-KT-003 validated): kotlinx.coroutines + dedicated reader OS
 * threads.** This is the Kotlin-native shape and the axis where the peer diverges from
 * Java (Java = platform/virtual threads; Kotlin = structured coroutines over a thread
 * pool). One **dedicated OS thread per connection** runs the blocking framed read loop —
 * the §7b transport rule: NEVER run a blocking syscall (`read`/`accept`) on a bounded
 * cooperative coroutine pool (it would starve `Dispatchers.Default`); blocking I/O lives
 * on dedicated threads, and the dispatched WORK runs as coroutines on
 * [Dispatchers.IO]. The reader demuxes inbound frames (§6.11): an EXECUTE_RESPONSE
 * routes to its awaiting outbound caller by request_id through a
 * `ConcurrentHashMap<requestId, CompletableDeferred<Envelope>>` correlation table (the
 * coroutine analogue of Java's CompletableFuture / OCaml's per-thread demux); an inbound
 * EXECUTE is dispatched in its OWN coroutine (§4.8) so a handler that originates an
 * outbound EXECUTE (§6.13(b)) and `await`s its response does NOT block the reader.
 * Writes are serialized by a per-connection write lock. The N7 demux is the 8-way
 * concurrent check in the smoke. The one non-stdlib runtime dep is kotlinx-coroutines
 * (profile [deps]); the sockets are stdlib `java.net`.
 */
object Transport {

    /** Per-connection IO: the framed stream, the write lock, and the §6.11 demux table. */
    class Io(private val socket: Socket) {
        private val input = DataInputStream(socket.getInputStream())
        private val out: OutputStream = BufferedOutputStream(socket.getOutputStream())
        private val writeLock = Any()
        // request_id → completion; the reader completes it with the correlated response.
        private val pending = ConcurrentHashMap<String, CompletableDeferred<Envelope?>>()

        @Volatile private var closed = false

        init {
            socket.tcpNoDelay = true // §7b: TCP_NODELAY on raw-socket peers
        }

        internal val dataInput: DataInputStream get() = input

        fun writeFramed(env: Envelope) {
            val payload = Wire.frameOfEnvelope(env)
            synchronized(writeLock) { Wire.writeFrame(out, payload) }
        }

        /** §6.13(b) outbound primitive: send a request envelope, suspend-await its
         *  correlated EXECUTE_RESPONSE (§6.11). The reader routes the response. Returns
         *  null if the connection closes first or the wait times out (§6.12). */
        suspend fun outbound(request: Envelope): Envelope? {
            val requestId = request.root.text("request_id") ?: ""
            val deferred = CompletableDeferred<Envelope?>()
            pending[requestId] = deferred
            try {
                writeFramed(request)
                if (closed) return null
                return withTimeoutOrNull(30_000) { deferred.await() }
            } catch (e: EntityTransportException) {
                return null
            } finally {
                pending.remove(requestId)
            }
        }

        internal fun routeResponse(env: Envelope) {
            val requestId = env.root.text("request_id") ?: ""
            pending.remove(requestId)?.complete(env)
        }

        fun close() {
            closed = true
            // wake any waiters so a reentrant outbound doesn't hang past close (→ null).
            for (d in pending.values) d.complete(null)
            try {
                socket.close()
            } catch (ignore: IOException) {
                // best-effort
            }
        }
    }

    /** A running listener: the bound port plus a handle to stop it. */
    class Listener(
        private val server: ServerSocket,
        val port: Int,
        private val scope: CoroutineScope,
        private val job: Job,
    ) : AutoCloseable {
        override fun close() {
            try {
                server.close()
            } catch (ignore: IOException) {
                // best-effort
            }
            job.cancel()
        }
    }

    /** The reader loop (§6.11 demux): EXECUTE_RESPONSE → route; EXECUTE → dispatch in its
     *  own coroutine (§4.8) + write the response. Runs on a dedicated OS thread (§7b: the
     *  blocking framed read never sits on the cooperative pool). Returns when the
     *  connection closes / a malformed frame ends it. */
    /**
     * §4.11's table (0.8.2.25): the `(status, code, message)` a pre-admission failure's
     * CAUSE takes.
     *
     * *"The frame obligation belongs to the class; the CODE belongs to the cause
     * `[MUST]`"* — a single code for the class would answer an honest caller under the
     * wrong reason and send them to the wrong layer.
     *
     * | cause | answer | stated at |
     * |---|---|---|
     * | connect-auth proof-of-possession | `401 authentication_failed` | §4.6/§4.7 — the connect handler's |
     * | envelope over the configured max | `413 payload_too_large` | §4.10(a), N14 |
     * | resolution integrity (mis-keyed `included`) | `400 hash_mismatch` | §5.2a, §1.8 |
     * | framing / never becomes an Envelope | `400 invalid_request` | §4.7, §4.11 |
     * | root is neither EXECUTE nor EXECUTE_RESPONSE | `400 invalid_request` | in the loop, not here |
     *
     * THE TAG ARM KEEPS `non_canonical_ecf` AND THAT IS DELIBERATE. §4.11 rules that code
     * non-conformant *"on the framing arm"* and gives its reason in the same sentence:
     * `ENTITY-CBOR-ENCODING` *"defines that code for CBOR tag-policy violations
     * specifically"*, which that document still MUSTs at decode time (§6.3). The two rows
     * are disjoint by CAUSE rather than in conflict: a tag in a DATA-FIELD position is the
     * policy violation with its own code, while a tag in the fixed envelope or
     * entity-wrapper maps is a structurally invalid frame — i.e. the framing arm. Everything
     * else this decoder calls non-canonical (a non-minimal head, an indefinite length,
     * mis-ordered keys) is genuinely "non-canonical CBOR that never becomes an Envelope" and
     * takes `invalid_request`.
     *
     * The messages are a FIXED TABLE, never the internal exception text: a wire-visible
     * string must stay ASCII (two peers in this cohort have been killed at runtime by a
     * non-ASCII byte in an encoded string, on two unrelated compilers), and nothing here
     * echoes attacker-supplied bytes back.
     */
    internal fun preAdmissionRefusal(e: Throwable): Triple<Int, String, String> = when {
        e is FrameTooLargeException ->
            Triple(413, "payload_too_large", "inbound frame exceeds the configured maximum size")
        e is HashMismatchException ->
            Triple(400, "hash_mismatch", "an entity was addressed by a hash that does not bind to it")
        e is CodecRefusalException && e.error is EntityError.CodecError.TagRejected ->
            Triple(400, "non_canonical_ecf", "CBOR tags are forbidden anywhere in an entity data field")
        else -> Triple(400, "invalid_request", "frame did not decode into an envelope")
    }

    /**
     * Recover ONLY the `request_id` from a frame the strict decoder rejected, so the refusal
     * can be delivered CORRELATED rather than as §4.11's uncorrelated best-effort frame.
     * `""` when nothing is recoverable.
     *
     * The frame stays rejected: nothing is built from it, nothing is stored, and a tag is
     * never interpreted — the salvage decode exists solely to read back the correlation key.
     * The envelope and entity-wrapper shapes are fixed maps with no legal tag position, so a
     * frame whose ONLY defect is a tag inside some entity's `data` still has a structurally
     * sound root, which is exactly the case worth recovering (and the one CAP-6a's `>2^64`
     * half arrives as — a bignum can only reach a peer as a major-type-6 tag).
     */
    private fun salvageRequestId(payload: ByteArray): String = try {
        val v = (CanonicalCbor.decodeSalvage(payload) as? EcfResult.Ok)?.value
        val root = Cbor.asMap((v as? EcfValue.MapVal)?.get("root"))
        (Cbor.asMap(root?.get("data"))?.get("request_id") as? EcfValue.Text)?.value ?: ""
    } catch (bad: Exception) {
        ""
    }

    /**
     * Put the coded EXECUTE_RESPONSE §4.11 (0.8.2.25) requires on the wire for a frame
     * refused BEFORE it becomes an admitted request.
     *
     * *"A peer that refuses a frame pre-admission MUST put a coded EXECUTE_RESPONSE on the
     * wire `[MUST]` — correlated by `request_id` where the id is available, and otherwise as
     * a best-effort coded frame carrying no correlation."*
     *
     * §4.9(c)'s deliver-or-signal rule is scoped to *"every request the peer ADMITS"* and
     * therefore reaches none of these, which is why §4.11 exists. Both of the non-conformant
     * behaviours it names separately were present on this peer: DROPPING the frame (the
     * un-salvageable decode arm and the non-EXECUTE root, *"the weaker of the two precisely
     * because nothing surfaces it"*) and CLOSING with no coded frame (the oversize and
     * truncated arms).
     *
     * AN EMPTY `requestId` IS THE BEST-EFFORT FORM, not a bug: it is what the section
     * prescribes where no id can be recovered.
     */
    private fun refusePreAdmission(io: Io, requestId: String, cause: Throwable) {
        val (status, code, message) = preAdmissionRefusal(cause)
        writeRefusal(io, requestId, status, code, message)
    }

    private fun writeRefusal(io: Io, requestId: String, status: Int, code: String, message: String) {
        try {
            io.writeFramed(Envelope(Wire.makeResponse(requestId, status, Wire.errorResult(code, message))))
        } catch (ignore: Exception) {
            // A write failure here is a dead socket, not a protocol decision; the read
            // loop's own error handling ends the connection on the next iteration.
        }
    }

    private fun readLoop(peer: Peer, conn: Conn, io: Io, scope: CoroutineScope) {
        try {
            while (true) {
                val payload = try {
                    Wire.readFrame(io.dataInput) ?: break // clean EOF
                } catch (refusal: EntityTransportException) {
                    if (refusal is FrameTooLargeException || refusal is TruncatedFrameException) {
                        // The stream is desynchronized on both REFUSABLE arms — an oversize
                        // body was never drained, a truncated one never arrived — so the
                        // coded frame goes out and THEN the connection closes. §4.11 makes
                        // the frame mandatory and leaves the close to us; closing is the
                        // only sound choice once the framing is lost, and it is a CHOICE
                        // rather than an alternative to answering. An ordinary hangup is not
                        // a refusal and gets nothing, which is what the EOF arm separates.
                        //
                        // §4.11's best-effort UNCORRELATED form: no request_id can be
                        // recovered from a frame whose body never arrived, and guessing one
                        // would correlate the refusal to somebody else's in-flight request.
                        refusePreAdmission(io, "", refusal)
                    }
                    break
                }
                val env = try {
                    Wire.envelopeOfFrame(payload)
                } catch (bad: Exception) {
                    // A COMPLETE frame the decoder refused. The framing is intact, so we
                    // answer and KEEP SERVING — and the refusal MUST be a STATUS, not
                    // silence (§4.11). Skipping it satisfies only the first half of the
                    // sentence and leaves the sender blocked until its own timeout, so a
                    // refusal is indistinguishable from a dead peer. §4.9(c) says the same
                    // from the other direction.
                    //
                    // THE CODE IS THE CAUSE'S (§4.11, §5.2a). This answered
                    // `non_canonical_ecf` for every cause until 0.8.2.24/.25 pinned them
                    // apart: a mis-keyed `included` entry is `400 hash_mismatch` (its
                    // encoding is canonical — what is false is the claim the key makes), a
                    // tag-policy violation keeps `non_canonical_ecf`, and everything else
                    // that never becomes an Envelope is `400 invalid_request`.
                    refusePreAdmission(io, salvageRequestId(payload), bad)
                    continue
                }
                if (env.root.type == "system/protocol/execute/response") {
                    io.routeResponse(env)
                } else if (env.root.type != "system/protocol/execute") {
                    // §6.5's "Other type?" arm, as rewritten at 0.8.2.25 (N12/N17): "400
                    // invalid_request, coded frame; MAY then close (§3.3, §4.11). NOT a bare
                    // close — that is indistinguishable from a network fault."
                    //
                    // §3.3 read "the connection MUST be closed", assigning no code and
                    // requiring no frame; this peer did something weaker still — `dispatch`
                    // answers null for a non-EXECUTE root and the loop wrote NOTHING, which
                    // is §4.11's silent-drop failure. §9.1's floor row that MANDATED the bare
                    // close was REPLACED at the same revision (N18).
                    //
                    // The request_id is read best-effort: an arbitrary root type is under no
                    // obligation to carry one, and §4.11 licenses the uncorrelated frame
                    // exactly there. We do NOT close — on a multiplexed connection that would
                    // cost every ADMITTED in-flight request its response.
                    writeRefusal(
                        io, env.root.text("request_id") ?: "", 400, "invalid_request",
                        "root entity is neither EXECUTE nor EXECUTE_RESPONSE",
                    )
                } else {
                    // §4.8 inbound concurrent with outbound: dispatch on its own coroutine
                    // (Dispatchers.IO) so a handler can reenter (§6.11) without blocking
                    // this reader.
                    scope.launch(Dispatchers.IO) {
                        val resp = try {
                            peer.dispatch(conn, env)
                        } catch (e: RuntimeException) {
                            Envelope(Wire.makeResponse(env.root.text("request_id") ?: "", 500,
                                Wire.errorResult("internal_error", null)))
                        }
                        if (resp != null) {
                            try {
                                io.writeFramed(resp)
                            } catch (ignore: EntityTransportException) {
                                // write failure ends this exchange; reader keeps going
                            }
                        }
                    }
                }
            }
        } catch (e: EntityTransportException) {
            // framing fault ends the connection
        } finally {
            io.close()
        }
    }

    /** Bind 127.0.0.1:port (0 = auto) and spawn the accept loop. */
    fun startListener(peer: Peer, port: Int): Listener {
        val server = ServerSocket()
        server.reuseAddress = true
        server.bind(InetSocketAddress(InetAddress.getLoopbackAddress(), port), 64)
        val bound = server.localPort
        val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
        // accept on a dedicated OS thread (blocking accept off the cooperative pool, §7b).
        val acceptThread = thread(name = "ec-accept", isDaemon = true) {
            while (!server.isClosed) {
                val client = try {
                    server.accept()
                } catch (e: IOException) {
                    break // socket closed → stop
                }
                serveConnection(peer, client, scope)
            }
        }
        val job = scope.launch { /* keep scope alive for the listener's lifetime */ }
        return Listener(server, bound, scope, job).also {
            // tie the accept thread to the job's cancellation.
            scope.coroutineContext[Job]!!.invokeOnCompletion { acceptThread.interrupt() }
        }
    }

    private fun serveConnection(peer: Peer, client: Socket, scope: CoroutineScope) {
        val io = try {
            Io(client)
        } catch (e: IOException) {
            try { client.close() } catch (ignore: IOException) {}
            return
        }
        val conn = Conn()
        // wire the §6.13(b) outbound seam to this connection (§6.11 reentry).
        conn.outbound = { env -> io.outbound(env) }
        // dedicated reader thread per connection (§7b: blocking read off the pool).
        thread(name = "ec-reader", isDaemon = true) { readLoop(peer, conn, io, scope) }
    }

    // ══════════════════════════════════════════════════════════════════════════════
    // Client side — the dialer + initiator handshake (drives the two-peer loopback)
    // ══════════════════════════════════════════════════════════════════════════════

    /** A dialed, authenticated session (§4.4): the IO, the minted cap + granter + sig. */
    class Session internal constructor(private val io: Io, private val local: Identity) : AutoCloseable {
        private val reqCounter = AtomicInteger()
        var remotePeerId: String? = null
            internal set
        var capability: Entity? = null
            internal set
        /** The remote peer identity that granted the session cap (the §4.4 granter). */
        var granterPeer: Entity? = null
            internal set
        /** The signature over the session cap (travels with it in `included`). */
        var capSignature: Entity? = null
            internal set

        internal fun nextRequestId(): String = "req-${reqCounter.incrementAndGet()}"

        /** Send REQUEST and await its correlated EXECUTE_RESPONSE (request_id demux). */
        suspend fun send(request: Envelope): Envelope? = io.outbound(request)

        /** Build, sign, and send an authenticated EXECUTE; await the response. The full
         *  §5.8 authority chain travels in `included`. */
        suspend fun execute(uri: String, operation: String, params: Entity, resource: EcfValue.MapVal?): Envelope? {
            val cap = capability!!
            val exec = Wire.makeExecute(nextRequestId(), uri, operation, params,
                local.identityHash(), cap.hash(), resource)
            val execSig = local.sign(exec)
            val inc = listOf(
                Envelope.Included(cap.hash(), cap),
                Envelope.Included(granterPeer!!.hash(), granterPeer!!),
                Envelope.Included(local.identityHash(), local.peerEntity),
                Envelope.Included(capSignature!!.hash(), capSignature!!),
                Envelope.Included(execSig.hash(), execSig),
            )
            return send(Envelope(exec, inc))
        }

        override fun close() {
            io.close()
        }
    }

    /** Open a client connection to host:port and start its reader thread, then drive the
     *  §4.1 forward handshake. Returns the authenticated session. */
    fun dial(initiator: Peer, host: String, port: Int): Session = runBlocking {
        val sock = try {
            Socket(host, port)
        } catch (e: IOException) {
            throw EntityTransportException("dial failed", e)
        }
        val io = Io(sock)
        val session = Session(io, initiator.identity)
        // the client reader: a core responder sends only EXECUTE_RESPONSEs; route them. A
        // reentrant inbound EXECUTE (§6.11) is dispatched on its own coroutine too.
        val conn = Conn()
        conn.outbound = { env -> io.outbound(env) }
        val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
        thread(name = "ec-client-reader", isDaemon = true) { readLoop(initiator, conn, io, scope) }
        handshake(session, initiator.identity)
        session
    }

    /** Drive the §4.1 forward handshake as initiator: hello then authenticate. On
     *  success, populate the session with the §4.4 capability the responder minted. */
    private suspend fun handshake(s: Session, local: Identity) {
        // ── hello ──
        val hello = Entity.make("system/protocol/connect/hello",
            Cbor.map(
                "peer_id", local.peerId,
                "nonce", Cbor.bytes(randomNonce()),
                "protocols", Cbor.textArray("entity-core/1.0"),
                "timestamp", EcfValue.IntVal.of(Capability.nowMs()),
                "hash_formats", Cbor.textArray("ecfv1-sha256"),
                "key_types", Cbor.textArray("ed25519"),
            ))
        val r1 = s.send(Envelope(Wire.makeExecute(s.nextRequestId(), "system/protocol/connect", "hello", hello)))
        requireOk(r1, "hello")
        val remoteHello = Wire.responseResult(r1!!)!!
        s.remotePeerId = remoteHello.text("peer_id")
        val remoteNonce = remoteHello.bytes("nonce")!!

        // ── authenticate ──
        val auth = Entity.make("system/protocol/connect/authenticate",
            Cbor.map(
                "peer_id", local.peerId,
                "public_key", Cbor.bytes(local.publicKey()),
                "key_type", "ed25519",
                "nonce", Cbor.bytes(remoteNonce),
            ))
        val authSig = local.sign(auth)
        val authInc = listOf(
            Envelope.Included(local.identityHash(), local.peerEntity),
            Envelope.Included(authSig.hash(), authSig),
        )
        val r2 = s.send(Envelope(
            Wire.makeExecute(s.nextRequestId(), "system/protocol/connect", "authenticate", auth), authInc))
        requireOk(r2, "authenticate")

        // parse the §4.4 initial capability grant
        val grant = Wire.responseResult(r2!!)!!
        val tokenH = grant.bytes("token")!!
        val token = r2.includedGet(tokenH)
            ?: throw EntityTransportException("authenticate grant omits the capability token")
        val granterH = token.bytes("granter")!!
        val granterPeer = r2.includedGet(granterH)
            ?: throw EntityTransportException("authenticate grant omits the granter identity")
        val capSig = Capability.findSignature(token.rawHash(), r2.included)
            ?: throw EntityTransportException("authenticate grant omits the capability signature")
        s.capability = token
        s.granterPeer = granterPeer
        s.capSignature = capSig
    }

    private val NONCE_RNG = java.security.SecureRandom()

    private fun randomNonce(): ByteArray {
        val b = ByteArray(32)
        NONCE_RNG.nextBytes(b)
        return b
    }

    private fun requireOk(env: Envelope?, step: String) {
        if (env == null) throw EntityTransportException("$step failed: no response")
        val status = Wire.responseStatus(env)
        if (status != 200) {
            val r = Wire.responseResult(env)
            val code = r?.text("code")
            val msg = r?.text("message")
            throw EntityTransportException("$step failed: $status $code ${msg ?: ""}")
        }
    }
}
